### PR TITLE
[MM-47391] Upgrade date picker to v8 and making it accessible

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Upgrading react-day-picker to v8 in order to use their accessibility features, this forced me to also install date-fns as it is a peer dependency. `react-focus-trap` is very useful for easily trapping focus on the date picker when using react popper.

Date picker for search, DND and custom statuses is now accessible by using the arrows keys when the picker appears.

This PR also includes a bunch of fixes to the custom statuses modal in order to make it accessible.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47391
https://mattermost.atlassian.net/browse/MM-47530

#### Release Note

```release-note
Making the dater picker keyboard accessible by upgrading react-day-picker to v8.3.6. Also introduced date-fns v2.29.3 because react-day-picker uses it as a peer dependency
```

### CODE REVIEW

**Review by cyrilzhang-mm — CHANGES_REQUESTED** [review]
This is definitely an improvement over what we have now. I've noticed a few things that we could do to make accessibility even better. 

I think we can (easily?) fix these:

1. Can you prevent this element from being focused? It seems like you have added the code to `GenericModal` to make this happen, but it's still being focused

Screen Shot 2022-11-18 at 9 25 01 AM

2. The emoji picker in the status modal currently no longer automatically focuses when the emoji button is clicked, making it impossible to use.

It would nice to be able to fix these, but they might be out of scope for now:

1. When the custom status modal is first opened, it's unclear where the focus is. Focusing the close button first would be the right thing to do, however a11y-controller might make this impossible. Maybe after merging my changes to the a11y-controller, we can do it.

2. Does the focus trap library have the ability to prevent focus from escaping to the address bar?

3. After selecting a duration from the duration dropdown in the status modal, it should re-focus the button that opens the duration dropdown. Currently it looks like the focus just disappears. Again, maybe another fix for after a11y-controller changes. 

4. When the date picker is open, there is no way to close it other than by picking a date. I would expect escape key to be able to close it, but in the status and DND modal, pressing escape closes the whole modal. Same for the time picker. Pressing escape key in the search bar doesn't do anything.

**Review by michelengelen — APPROVED** [review]
lgtm

**Review by nevyangelova — APPROVED** [review]
LGTM. Need to update snapshots 👍

**Comment by BenCookie95** [comment]
@nevyangelova & @michelengelen I made some more changes that need to be re-reviewed, they're all in here https://github.com/mattermost/mattermost-webapp/pull/11609/commits/cadefbe7221f326d8d686a752c22627601dd909d.

Main change is allowing the user to close the datepicker with the escape key

File: components/date_picker/date_picker.tsx
Comment (michelengelen) [inline components/date_picker/date_picker.tsx:28]: nit:

```suggestion
    const [popperElement, setPopperElement] = useState<HTMLDivElement | null>(null);
```

File: components/custom_status/custom_status_modal.tsx
Comment (cyrilzhang-mm) [inline components/custom_status/custom_status_modal.tsx:212]: Here, too.

File: components/date_picker/date_picker.tsx
Comment (BenCookie95) [reply components/date_picker/date_picker.tsx:52]: The reason to import them like this is because we need to dynamically import locales because we don't know what locale the user has beforehand. There is no way to import all locales at once and unfortunately we can't use moment with this package

File: components/date_picker/date_picker.tsx
Comment (BenCookie95) [reply components/date_picker/date_picker.tsx:10]: We can't use moment to get the locales for the date picker package because there are some discrepancies between moment and date-fns

File: components/custom_status/custom_status.scss
Comment (BenCookie95) [reply components/custom_status/custom_status.scss:70]: Yeah, it's overriding another style from somewhere else in order to have this input match the other styles in the modal

File: components/custom_status/custom_status.scss
Comment (nevyangelova) [reply components/custom_status/custom_status.scss:88]: yes lets avoid !importants and styling by html elements. Perhaps ass a class to the input?

File: components/custom_status/custom_status.scss
Comment (nevyangelova) [reply components/custom_status/custom_status.scss:82]: Same here

File: components/date_picker/date_picker.tsx
Comment (nevyangelova) [reply components/date_picker/date_picker.tsx:6]: We are avoiding adding new libraries for now, unless its a unicorn library that does exactly what we need, and we are even getting rid of a lot currently. So let's try to use a custom solution here if possible. Sorry for not catching this earlier!

Comment (nevyangelova) [reply components/date_picker/date_picker.tsx:6]: We are avoiding adding new libraries for now, unless its a unicorn library that does exactly what we need, and we are even getting rid of a lot currently. So let's try to use a custom solution here if possible. Sorry for not catching this earlier!

File: components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.scss
Comment (nevyangelova) [reply components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.scss:140]: +1

File: components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.tsx
Comment (michelengelen) [reply components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.tsx:186]: nit: naming it `togglePopper` is misleading, since you need to pass a parameter

File: components/date_picker/date_picker.tsx
Comment (hmhealey) [reply components/date_picker/date_picker.tsx:6]: I'm surprised that's not part of the popper library itself. I'm pretty sure even our old version of React Bootstrap handles that itself.

It looks like Floating UI has support for this via its `FloatingFocusManager`, so maybe we should switch this to use Floating UI instead of React Popper?

Comment (hmhealey) [reply components/date_picker/date_picker.tsx:6]: Actually, yeah, let's use Floating UI here. We actually only have 1 other place where we use react-popper, so let's avoid spreading that to keep the number of those libraries down

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx
Comment (hmhealey) [reply components/suggestion/search_date_suggestion/search_date_suggestion.tsx:8]: I swear `react-day-picker` is part of why `moment` crept its way into our code base to begin with, so that change is kind of annoying. I agree that it might be nice to look for a different library for this at some point, but at the very least, because we're dynamically loading these locales, I don't think this should impact bundle size until these components are actually used.

I'd keep an eye on this, but I think it's probably okay.

File: components/date_picker/date_picker.tsx
Comment (hmhealey) [inline components/date_picker/date_picker.tsx:24]: Making this a variable inside the functional component means that it won't be persisted between runs of this, so it'll have to be a global variable like the one in `suggestion_date_picker`. Since we're doing the same thing in both places, perhaps it would be better to pull this out into a reusable utility function of some sort? I'd say a hook, but I don't want to send us down the rabbit hole of trying to make the `Suggestion` components functional right now 😛

File: components/custom_status/date_time_input.tsx
Comment (hmhealey) [reply components/custom_status/date_time_input.tsx:130]: If Luxon can do that, that'd be preferred. We don't have a specific ticket or epic for it yet, but we're trying to move towards Luxon instead of Moment

File: components/date_picker/date_picker.scss
Comment (hmhealey) [reply components/date_picker/date_picker.scss:3]: You can also just use the non-rgb version
```suggestion
    background: var(--center-channel-bg);
```

**Comment by hmhealey** [comment]
tl;dr version of my comments since I didn't submit them in a single review for some reason:
- Adding `date-fns` isn't ideal, but it's probably fine since it's lazily loaded anyway. We can look at replacing the date picker at some point in the future, but I don't think it's blocking.
- For `focus-trap-react`, we can probably avoid adding this if we use one of the other popper libraries that we have. The Floating UI one seems like it should support this, and that'd be a better one to use here anyway because I'm pretty sure that that one was considered the best option when we last discussed those libraries.

File: components/custom_status/date_time_input.tsx
Comment (BenCookie95) [reply components/custom_status/date_time_input.tsx:126]: Using Luxon now, also replaced this function in the dnd modal

File: components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.scss
Comment (BenCookie95) [reply components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.scss:150]: Yeah the issue lies in the custom status file, will fix that in my next commit

File: components/suggestion/suggestion_box/suggestion_box.jsx
Comment (BenCookie95) [reply components/suggestion/suggestion_box/suggestion_box.jsx]: Added a guard and now using the accessibility controller

File: components/date_picker/date_picker.tsx
Comment (BenCookie95) [reply components/date_picker/date_picker.tsx]: Moved to a utility function in utils.tsx

**Comment by BenCookie95** [comment]
I think I have addressed all the concerns that were raised, please take another look!

**Review by cyrilzhang-mm — APPROVED** [review]
Good work, it's a big improvement for sure. Only one nit: Can we add `aria-haspopup='menu'` to the date and time buttons? 

Let's make tickets for the outstanding issues that can't be resolved in the PR, such as 
1. All modals don't trap focus correctly
2. Most dropdowns (across the whole app) don't announce whether they are open or not, and can't be used with arrow keys
3. We can't close most dropdowns in modals without closing the whole modal

**Comment by BenCookie95** [comment]
> Good work, it's a big improvement for sure. Only one nit: Can we add `aria-haspopup='menu'` to the date and time buttons?
> 
> Let's make tickets for the outstanding issues that can't be resolved in the PR, such as
> 
> 1. All modals don't trap focus correctly
> 2. Most dropdowns (across the whole app) don't announce whether they are open or not, and can't be used with arrow keys
> 3. We can't close most dropdowns in modals without closing the whole modal

@cyrilzhang-mm I guess my only hesitation of adding the `aria-haspopup='menu'` to the time button is that it would indicate to the user that they should be able to use the arrow keys to navigate but that isn't actually possible

**Comment by cyrilzhang-mm** [comment]
@BenCookie95 That's true. We can hold off on it until we improve the time picker.

**Review by nevyangelova — APPROVED** [review]

File: components/custom_status/__snapshots__/expiry_menu.test.tsx.snap
Comment (michelengelen) [inline components/custom_status/__snapshots__/expiry_menu.test.tsx.snap:24]: can we make this single line please?

```suggestion
          <button className="style--none expiry-wrapper expiry-selector">
```

File: components/custom_status/__snapshots__/expiry_menu.test.tsx.snap
Comment (michelengelen) [inline components/custom_status/__snapshots__/expiry_menu.test.tsx.snap:101]: same here

File: components/custom_status/custom_status_modal.tsx
Comment (michelengelen) [inline components/custom_status/custom_status_modal.tsx:134]: `handleKeyDown` is missing in the dependencies array

File: components/custom_status/custom_status_modal.tsx
Comment (michelengelen) [inline components/custom_status/custom_status_modal.tsx:140]: `props.onExited` is missing in the dependencies array

File: components/custom_status/date_time_input.tsx
Comment (michelengelen) [inline components/custom_status/date_time_input.tsx:71]: `handleKeyDown` is missing in the dependencies array

File: components/custom_status/date_time_input.tsx
Comment (michelengelen) [inline components/custom_status/date_time_input.tsx:77]: `handlePopperOpenState` is missing in the dependencies array

File: components/custom_status/expiry_menu.tsx
Comment (michelengelen) [inline components/custom_status/expiry_menu.tsx:117]: please remove the blank line

**Review by michelengelen — CHANGES_REQUESTED** [review]

File: components/custom_status/__snapshots__/expiry_menu.test.tsx.snap
Comment (BenCookie95) [reply components/custom_status/__snapshots__/expiry_menu.test.tsx.snap:24]: This is a generated snapshot, I can't update the layout because it will get overwritten

File: components/custom_status/__snapshots__/expiry_menu.test.tsx.snap
Comment (BenCookie95) [reply components/custom_status/__snapshots__/expiry_menu.test.tsx.snap:101]: Same as above

**Comment by BenCookie95** [comment]
@michelengelen & @hmhealey this is ready for re-review btw

File: components/custom_status/date_time_input.tsx
Comment (michelengelen) [inline components/custom_status/date_time_input.tsx:128]: nit: 
```suggestion
    const formatDate = (date: Date): string => DateTime.fromJSDate(date).toFormat('yyyy-MM-dd');
```

**Review by michelengelen — APPROVED** [review]
just one nit ... LGTM

**Comment by BenCookie95** [comment]
@matthewbirtch I've made most of the style changes you requested. You can inspect it in the DND modal but you can also see it in the custom status modal.

The search bar date picker is a little different though, the box has a set width so I have to stretch the date picker in order for it to fit. If you type `before:` you can inspect it. Let me know what I can do with it

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Looking great. A few minor adjustments @BenCookie95:

For the date numbers in the calendar:
- can we adjust to`font-weigh: 600`?
- For the dates that are enabled, can we adjust the opacity of the numbers to 100% instead of 72%? I'd like to differentiate more from the disabled dates.
- For the search date picker, today's date doesn't seem to be getting the background highlighted, but the number is getting bolded. Is that intentional?
    image

- Also in the search date picker, seems like there might be extra padding on the left/right/bottom. I had a hard time isolating which CSS rules to affect this, but these two elements seem to show it.
    image
    image

- For the `<` and `>` buttons at the top-right, can we add 8px of space between them?
- For the selected state, when I hover over it, it loses the selected state. Can we adjust this so that the selected date keeps the selected/active state visually while hovering?
- I'm not sure if it's caused by anything in this PR, but this popover seems to stay open even after I blur focus on the global search input
    image

**Comment by hmhealey** [comment]
Removing my review request since we have two other dev reviewers already

**Review by matthewbirtch — APPROVED** [review]
I think we're good to go with this @BenCookie95. I'm not going to worry about the extra padding on the search date picker. Thanks for the work on this!

**Review by furqanmlk — COMMENTED** [review]
Hi @BenCookie95 , Calendar dialogue doesn't close when clicking outside the dialogue. In Master it gets closed.

**Comment by BenCookie95** [comment]
@furqanmlk You will have to test this locally as the cloud test server was unable to pick up the latest changes. Should be working as expected now

**Review by furqanmlk — APPROVED** [review]
Thanks @BenCookie95, working fine

**Comment by AshishDhama** [comment]
@BenCookie95 I have a patch here for changes required for post reminder. I can also push it directly to this pr.
[post-reminder.patch](https://github.com/mattermost/mattermost-webapp/files/10858129/post-reminder.patch)

**Comment by BenCookie95** [comment]
> @BenCookie95 I have a patch here for changes required for post reminder. I can also push it directly to this pr. [post-reminder.patch](https://github.com/mattermost/mattermost-webapp/files/10858129/post-reminder.patch)

@AshishDhama Are there change in a pull request yet? This PR is already approved so don't want to push any more code changes

**Comment by BenCookie95** [comment]
e2e tests are passing here https://github.com/mattermost/mattermost-webapp/pull/12295

**Pull request merged**
